services/src/main/scala/cromwell/pipeline/service/GitLabProjectVersioning.scala

```diff
@@ -13,38 +13,73 @@ import scala.concurrent.{ ExecutionContext, Future }
 class GitLabProjectVersioning(httpClient: HttpClient, config: GitLabConfig)
     extends ProjectVersioning[VersioningException] {
 
-  override def updateFile(project: Project, projectFile: ProjectFile, version: Option[Version])(
+  override def updateFile(project: Project, projectFile: ProjectFile, version: Option[PipelineVersion])(
     implicit ec: ExecutionContext
   ): AsyncResult[String] = {
     val path = URLEncoder.encode(projectFile.path.toString, "UTF-8")
-    val fileUrl = s"${config.url}projects/${project.repository}/repository/files/$path"
-    val versionValue = version.map(_.name).getOrElse(config.defaultFileVersion)
+    val repositoryId: Repository = project.repository match {
+      case Some(repository) => repository
+      case None             => throw VersioningException(s"No repository for project: $project")
+    }
+    val fileUrl = s"${config.url}projects/${repositoryId.value}/repository/files/$path"
+
+    getProjectVersions(project).flatMap(versions => {
+      val latestVersion: Either[VersioningException, PipelineVersion] = (version, versions) match {
+        case (Some(_), Left(error)) => Left(error)
+        case (Some(userVersion), Seq(headVersion: GitLabVersion, _)) if headVersion.name >= userVersion =>
+          Left(
+            VersioningException(
+              s"Your version $userVersion is out of date. Current version of project: ${headVersion.name}"
+            )
+          )
+        case (Some(userVersion), Right(_))    => Right(userVersion)
+        case (None, Seq(v: GitLabVersion, _)) => Right(v.name)
+        case (None, Right(Seq()))             => Right(PipelineVersion(config.defaultFileVersion))
+        case (None, Left(error))              => Left(error)
+      }
+      latestVersion match {
+        case Left(error) =>
+          Future.successful(Left(error))
+        case Right(newVersion) =>
+          val payload =
+            Json.stringify(
+              Json.toJson(UpdateFileRequest(projectFile.content, newVersion.toString, config.defaultBranch))
+            )
+          httpClient.put(fileUrl, payload = payload, headers = config.token).flatMap {
+            case Response(HttpStatusCodes.OK, body, _) =>
+              createTag(repositoryId, newVersion).map {
+                case Right(_)        => Right(body)
+                case Left(exception) => Left(exception)
+              }
+            case _ =>
+              httpClient.post(fileUrl, payload = payload, headers = config.token).flatMap {
+                case Response(HttpStatusCodes.OK, body, _) =>
+                  createTag(repositoryId, newVersion).map {
+                    case Right(_)        => Right(body)
+                    case Left(exception) => Left(exception)
+                  }
+                case Response(_, body, _) => Future.successful(Left(VersioningException(body)))
+              }
+          }
+      }
+    })
```

Comment:
There are so many things that happen inside a single method. Maybe you should separate it into smaller methods like `validateVersion`, `sendUpdateVersionRequest`, or something like that.